NOISSUE - Add Event Sourcing Package

rodneyosodo opened the pull request:
Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

### What does this do?
Adds a seperate package to handle event sourcing just like messaging

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
- move event sourcing publisher to one place
- move event sourcing subscriber to one place
- add handler function to services that consume event such as bootstrap

### Have you included tests for your changes?
Yes

### Did you document any new/modified functionality?
Not yet

### Notes
The new es handler looks like this 
```go
type Event interface {
	Encode() (map[string]interface{}, error)
}

type Publisher interface {
	Publish(ctx context.Context, event Event) error
	Close() error
}

type EventHandler interface {
	Handle(ctx context.Context, event Event) error
}

type Subscriber interface {
	Subscribe(ctx context.Context, handler EventHandler) error
	Close() error
}

```

rodneyosodo committed:
Add event sourcing package

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Move producer to es package

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Use redis URL to configure username, password and db

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Change the redis subscriber interface

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Clean up publisher inaterface

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Update redis version

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Add tests

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Rename factory functions of events store

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Rename redis package to events package

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Remove unnecessary alias on redis event store

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Minor cosmetic changes

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Remove stream length from factory function

Set the default stream length to 1e9 in the events package. This is
because the stream length is not a required parameter in the factory
function. This commit also removes the stream length from the factory
function.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Remove group Name on subscribing

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

SammyOina commented on pkg/events/redis/publisher.go:
```diff
@@ -0,0 +1,107 @@
+		stream:            stream,
+	}
+
+	go es.StartPublishingRoutine(ctx)
```
I don't think we need to have StartPublishingRoutine exported

SammyOina commented on pkg/events/events.go:
```diff
@@ -0,0 +1,50 @@
+
+	// StartPublishingRoutine starts a routine that checks for unpublished events
+	// and publishes them to stream.
+	StartPublishingRoutine(ctx context.Context)
```
this doesn't need to be part of the interface or exported

rodneyosodo replied:
It is exported so that the different `es` should implement it. As we want the different es for example `Redis streams` or `NATS JS` to implement it as a functionality.

SammyOina replied:
do both of them require such a routine with the same signature and some possible es we might use in the future

rodneyosodo replied:
RabbitMQ and Redis don't support using buffers. Hence the implementation for Nats is ignore https://github.com/rodneyosodo/mainflux/blob/d140b10a4936bee0afbadd0c34486e70f5f4f153/pkg/events/nats/publisher.go#L77

rodneyosodo replied:
The function only takes in context for cancellation purposes as the implementation might vary.

dborovcanin replied:
Since it's purely related to the implementation, let's remove this method from the Publisher interface for the time being.

rodneyosodo committed:
Refactor package consumer to reflect changes in service name

The package consumer in the bootstrap/events/consumer/doc.go file has been modified to reflect the changes in the service name. The service name has been updated from "Things" to "Bootstrap". This change ensures that the package accurately represents the events consumer for the Bootstrap service.

This commit modifies the package consumer in the bootstrap/events/consumer/doc.go file to update the service name from "Things" to "Bootstrap". This change ensures that the package accurately represents the events consumer for the Bootstrap service.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Refactor main to remove go routines

Remove go routines from main when subscribing to events. This is because
the event handlers are already running in their own go routines.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Move cache to seperate package

This commit moves the cache package to a seperate package. This is
because the cache package is not the same as events package. The cache
package is used to store the data in memory while the events package is
used to send events to the event bus.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Make startPublishingRoutine private

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

drasko reviewed (approved):
LGTM
